perf-tool/src/agent.cpp

```diff
@@ -0,0 +1,154 @@
+ *******************************************************************************/
+
+#include <jvmti.h>
+#include <iostream>
+#include <stdio.h>
+#include <string.h>
+#include <sys/types.h>
+#include <thread>
+#include <unistd.h>
+
+#include "agentOptions.hpp"
+#include "infra.hpp"
+#include "json.hpp"
+#include "methodEntry.hpp"
+#include "monitor.hpp"
+#include "objectalloc.hpp"
+#include "server.hpp"
+#include "exception.hpp"
+
+using json = nlohmann::json;
+
+jvmtiEnv *jvmti;
+int portNo;
+std::string commandsPath = "";
+std::string logPath = "logs.json";
+
+JNIEXPORT jint JNICALL Agent_OnLoad(JavaVM *jvm, char *options, void *reserved)
+{
+    std::string token;
+    std::string optionsDelim = ",";
+    std::string pathDelim = ":";
+    std::string oIn = (std::string)options;
+    int pos1, pos2 = 0;
+    portNo = 9002;
+
+    // there is a max of two options the user can supply here
+    // "commands" is followed by a path to the commands file
+    // "log" is followed by the path to the location for the log file
+    // "portno" is followed by a number indicating the port to run the server on
+    while ((pos1 = oIn.find(optionsDelim)) != std::string::npos)
+    {
+        if (pos1 != std::string::npos)
+        {
+            token = oIn.substr(0, pos1);
+            if ((pos2 = token.find(pathDelim)) != std::string::npos)
+            {
+                if (!token.substr(0, pos2).compare("commandFile"))
+                {
+                    token.erase(0, pos2 + pathDelim.length());
+                    commandsPath = token;
+                }
+                else if (!token.substr(0, pos2).compare("logFile"))
+                {
+                    token.erase(0, pos2 + pathDelim.length());
+                    logPath = token;
+                }
+                else if (!token.substr(0, pos2).compare("portNo"))
+                {
+                    token.erase(0, pos2 + pathDelim.length());
+                    portNo = stoi(token);
+                }
+            }
+            oIn.erase(0, pos1 + optionsDelim.length());
+        }
+    }
+    if (!oIn.empty())
+    {
+        token = oIn;
+        if ((pos2 = token.find(pathDelim)) != std::string::npos)
+        {
+            if (!token.substr(0, pos2).compare("commandFile"))
+            {
+                token.erase(0, pos2 + pathDelim.length());
+                commandsPath = token;
+            }
+            else if (!token.substr(0, pos2).compare("logFile"))
+            {
+                token.erase(0, pos2 + pathDelim.length());
+                logPath = token;
+            }
+            else if (!token.substr(0, pos2).compare("portNo"))
+            {
+                token.erase(0, pos2 + pathDelim.length());
+                portNo = stoi(token);
+            }
+        }
+    }
+
+    std::cout << commandsPath << std::endl;
+    std::cout << logPath << std::endl;
+    std::cout << portNo << std::endl;
```

Comment:
a little bit of inconsistency, here you use `std::cout` and below you use `printf` to print to stdout. Chose one, and the preferred is `printf`.

Reply:
Question out of curiosity; why is the preferred `printf` if this is all C++ code?

Reply:
I was thinking along the lines of consistency. In some parts of the code there were `std::cout` while others there were `printf`, so I thought it would be better to have only one of them

Reply:
No I agree with you on the consistency bit; I just wanted to understand why `printf` over `cout`.

Reply:
Given that this is all C++ code there's not really a preference over one or the other. I think that would more of a personal preference in this case; however, I would lean towards `printf` because you need to explicitly define the types of params that we're printing to stdout which, at least for me, it makes it easier to read

Reply:
Ah I see; well given that this is all new C++ code, I would've preferred `cout` simply for consistency with C++ paradigms (unless there was a noticeable perf degradation, which is apparently something people complain about `cout`); also `printf` does allow one just lie about the type you're printing and one wouldn't realize the mistake until much later. That said, I'm not suggesting we go and change all the places now that the code's already in the repo; was just satisfying my curiosity is all :).

Reply:
That is true, I agree with your point of view. There's quite few aspects to consider when thinking about "to `cout` or to `printf`" besides personal preference that can slip through :)